### ISSUE TITLE
All-padding attention mask produces NaN softmax that propagates silently into encoder output

### ISSUE THREAD

**Issue by forkwright** [forkwright/logismos#41]
Migrated from `forkwright/logismos-archive#68`, because the archived tracker is not watched. The defect was re-verified against live `main` and every `file:line` below was re-derived from this repo, not copied from the archive. (`crates/transformers/src/attention.rs` differs from the archived tree only by a `clippy::expect_used` attribute inside the test module; the forward path cited below is unchanged.)

## Finding

When the attention `mask` is all zeros (every token masked), `mask_additive_in_place` sets every column of every score row to `-inf`. `softmax_last_dim` then computes `exp(-inf) = 0` for every entry, the row sum is `0`, and the normalizing division yields `NaN`.

The `NaN` is written back into `scores` and flows unchecked through the `scores @ v` matmul, `merge_heads`, and the output projection, so `forward()` returns a `Vec<f32>` full of `NaN` wrapped in `Ok(...)`. No guard exists between mask application and the softmax. An all-padding row is a valid input for a batch item that is entirely padding.

## Evidence

`crates/transformers/src/attention.rs:242-247` — the mask is applied unconditionally, per head:

```rust
        for h in 0..n_h {
            let start = h * seq * seq;
            let end = (h + 1) * seq * seq;
            let head = checked_slice_mut(&mut scores, start, end, "attention masked score head");
            cpu_f32::mask_additive_in_place(head, &col_mask, seq, seq);
        }
```

`crates/transformers/src/attention.rs:250-256` — the softmax is then taken with no check that any score survived masking:

```rust
        for h in 0..n_h {
            let start = h * seq * seq;
            let end = (h + 1) * seq * seq;
            let head = checked_slice_mut(&mut scores, start, end, "attention scores head");
            let sm = cpu_f32::softmax_last_dim(head, seq, seq);
            head.copy_from_slice(&sm);
        }
```

The poisoned `scores` then feed the matmul at `crates/transformers/src/attention.rs:268`, `merge_heads` at `crates/transformers/src/attention.rs:274`, and are returned as `Ok(out)` at `crates/transformers/src/attention.rs:278`.

The NaN-producing step itself is in the kernel: `crates/kernels/src/cpu_f32.rs:232` initialises the running max to `f32::NEG_INFINITY`, and `crates/kernels/src/cpu_f32.rs:246` inverts a `NaN` accumulator with no guard. That kernel-level defect is tracked separately; this issue tracks the caller's missing guard, and either fix alone closes the `Ok(NaN)` path.

## Why this matters

The transformer encoder feeds on-device embedding and memory pipelines that run without a network oracle to sanity-check results. A single all-padding batch item silently poisons the embedding it produces; the `NaN` surfaces only at some downstream consumer that happens to test for it, far from the source, where it is hard to attribute.

An adversary able to influence batch composition — or any malformed input that pads an entire item — can use this to silently corrupt or denial-of-service the embedding stage without raising an error, because the call reports success.

## Desired correction

Before calling `softmax_last_dim`, detect a fully-masked row (equivalently, validate that the mask for the row has at least one non-zero entry) and either return `Err(Error::Shape("attention mask is fully masked — no real tokens".into()))` or substitute an explicit uniform distribution for that row. Do not allow `NaN` to be returned inside `Ok`.

Done when: `forward()` called with `mask = &[0u8; seq]` returns `Err` (or a documented uniform-distribution output) instead of `Ok` containing `NaN`.

**Comment by forkwright** [comment]
Verified against live `main` (`f4999108040057ab0998143b54078d6640bd5c6`, current tip as of this comment) rather than the archive copy. **This defect is already fixed and closed by a mislinked PR** — same shape as forkwright/logismos#35/#36: the fixing PR used `Refs #41` instead of `Closes #41`, so GitHub never auto-closed this issue.

## What already fixed it

PR #66 (`fix(kernels): guard softmax_last_dim against a fully-masked row's NaN`, merged 2026-08-09T00:37:23Z) fixed the *kernel-level* NaN this issue's own "Evidence" section named as the actual source (`cpu_f32.rs:232`/`:246`), and the issue's own text says: *"either fix alone closes the `Ok(NaN)` path."* The kernel fix landed, and per that issue's own logic, closes this one too.

## Confirmed live on main, not just inferred from PR #66's description

1. **The kernel guard is in place.** `crates/kernels/src/cpu_f32/mod.rs:266-287` (moved from the single-file `cpu_f32.rs` by #88's later restructure, content unchanged) — `softmax_last_dim` detects a row whose max is `NEG_INFINITY` (a fully-masked row, exactly what an all-zero `mask` produces via `mask_additive_in_place`) and substitutes a uniform `1.0/n` distribution instead of computing `NaN.exp()`. Comment cites `forkwright/logismos#30`, the companion kernel-level issue.
2. **The caller path (this issue's actual ask) is directly tested.** `crates/transformers/src/attention.rs:458-472`, test `forward_with_fully_masked_row_returns_finite_output_not_nan`, explicitly cites `forkwright/logismos#41` in its WHY comment, calls `attention.forward(..., mask=&[0], ...)` (the exact `Done when` scenario) and asserts `out.iter().all(|v| v.is_finite())`.
3. **That test runs the shipped code and passes in CI, not a standalone reimplementation.** CI run 31922952513 (job `full-gate-build`, workflow `Gate Attestation`, `pull_request` event, conclusion `success`, this is a public repo so minutes are unmetered):
   ```
   PASS [0.006s] (200/212) transformers attention::tests::forward_with_fully_masked_row_returns_finite_output_not_nan
   ```
   https://github.com/forkwright/logismos/actions/runs/31922952513

## Done-when check

> `forward()` called with `mask = &[0u8; seq]` returns `Err` (or a documented uniform-distribution output) instead of `Ok` containing `NaN`.

Met by option B (documented uniform-distribution output) — confirmed by the test above, on `main`, in CI, today.

No code change needed for this issue. Closing with this evidence rather than opening a redundant PR.
